### ISSUE TITLE
no-inferred-empty-object-type not detecting all generics

### ISSUE THREAD

**Issue by OliverJAsh** [palantir/tslint#3416]
### Bug Report

- __TSLint version__: 5.8.0
- __TypeScript version__: 2.5.3
- __Running TSLint via__: (pick one) CLI

#### TypeScript code being linted

```ts
    {
        const wrapper = <L, A>(l: L) => ({ l });
        const x = wrapper(1); // expected error, no error
        // The second generic, A, is inferred as an empty object
    }
```

with `tslint.json` configuration:

```json
{
    "rules": {
        "no-inferred-empty-object-type": true
    }
}
```

/cc @rsxdalv

**Comment by ajafff** [comment]
Is this just an overly simplified example? `A` is never used anywhere

**Comment by OliverJAsh** [comment]
@ajafff That's true, however I still want tslint to tell me about the inferred empty object, even if the generic is unused in the function definition.

The generics could be used in the return type, but tslint still does not error:

``` ts
    class Wrapper<L, A> {
        _L: L;
        _A: A;
    }
    const wrapper = <L, A>(l: L): Wrapper<L, A> => new Wrapper();
    const x = wrapper(1); // expected error, no error
```

For context, an example of a real world function like this is the `Either` type https://github.com/gcanti/fp-ts/blob/c2caee7ca5da32651c0d106d2232d76f9d15a133/src/Either.ts#L301

**Comment by rsxdalv** [comment]
@OliverJAsh There's definitely bugs in this one.

To me it seems that return type enables this rule to function, so this one does error (however, isn't useful):
```ts
const wrapper = <L, A>(l: A): L => ({ l });
const x = wrapper(1); 
```

**Comment by ajafff** [comment]
I have it working in a branch https://github.com/ajafff/tslint/commit/9635993f5a7bb0e9b2b97ec3e49c6a914de3d5de

Unfortunately this makes it a bit too strict IMO. If an optional parameter is not passed, it's TypeParameter will be inferred as empty object.
Consider the following code:

```ts
declare let fn<T>(obj?: T, cb: (obj: T) => void): void;

fn(); // error here, because T is inferred as {}
```

**Comment by OliverJAsh** [comment]
@ajafff Personally that is desired behaviour. I want to know when empty object is inferred regardless of how.

**Comment by ajafff** [comment]
I polished up my implementation and submitted #3445 to fix this issue

**Comment by JoshuaKGoldberg** [comment]
### 💀 _**It's time!**_ 💀

**TSLint is deprecated** and no longer accepting pull requests other than security fixes. See #4534. ☠️ 
We recommend you instead use **[typescript-eslint](https://typescript-eslint.io)** to lint your TypeScript code with **ESLint**. ✅

👋 It was a pleasure open sourcing with you!

**Comment by JoshuaKGoldberg** [comment]
🤖 Beep boop! 👉  **TSLint is deprecated** 👈 _(#4534)_  and you should **switch to [typescript-eslint](https://typescript-eslint.io)**! 🤖 

🔒 This issue is being locked to prevent further unnecessary discussions. Thank you! 👋
